In [1]:
import pandas as pd
from sklearn.model_selection import StratifiedKFold
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score
import statistics
from sklearn.metrics import confusion_matrix

In [22]:

path='spambase.data'

columns = ['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d', 'word_freq_our', 'word_freq_over', 'word_freq_remove', 'word_freq_internet', 'word_freq_order', 'word_freq_mail', 'word_freq_receive', 'word_freq_will', 'word_freq_people', 'word_freq_report', 'word_freq_addresses', 'word_freq_free', 'word_freq_business', 'word_freq_email', 'word_freq_you', 'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000', 'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george', 'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet', 'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85', 'word_freq_technology', 'word_freq_1999', 'word_freq_parts', 'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting', 'word_freq_original', 'word_freq_project', 'word_freq_re', 'word_freq_edu', 'word_freq_table', 'word_freq_conference', 'char_freq_;', 'char_freq_(', 'char_freq_[', 'char_freq_!', 'char_freq_$', 'char_freq_#', 'capital_run_length_average', 'capital_run_length_longest', 'capital_run_length_total', 'spam']
data = pd.read_csv(path,  delimiter=",",names=columns)
X = data.iloc[:,0:57]
y = data['spam']

In [3]:
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...   char_freq_;  char_freq_(  \
0             0.00            0.00  ...          0.00        0.000   
1             0.00            0.94  ...          0.00        0.132   
2             0.64            0.25  ...          0.01        0.143   
3             0.31            0.63  ...          0.00        0.137   
4             0.31            0.63  ...          0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [4]:
clf_NB = GaussianNB()
clf_DT = DecisionTreeClassifier()
clf_LR = LogisticRegression()

In [5]:
skf = StratifiedKFold(n_splits=10)
skf.get_n_splits(X,y)

10

In [6]:
# # Sensitivity, hit rate, recall, or true positive rate
# TPR = TP/(TP+FN)
# # Specificity or true negative rate
# TNR = TN/(TN+FP) 
# # Precision or positive predictive value
# PPV = TP/(TP+FP)
# # Negative predictive value
# NPV = TN/(TN+FN)
# # Fall out or false positive rate
# FPR = FP/(FP+TN)
# # False negative rate
# FNR = FN/(TP+FN)
# # False discovery rate
# FDR = FP/(TP+FP)

# # Overall accuracy
# ACC = (TP+TN)/(TP+FP+FN+TN)

# TP|FP
# FN|TN

In [9]:
fpr_NB=[]
fnr_NB=[]
ter_NB=[]
acc_NB=[]
for train_index, test_index in skf.split(X,y):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y[train_index], y[test_index]
    
    clf_NB.fit(X_train, y_train)
    pred = clf_NB.predict(X_test)
    
    acc = round(accuracy_score(y_test, pred),4)
    a=confusion_matrix(y_test, pred)
    tp=a[0][0]
    fp=a[0][1]
    fn=a[1][0]
    tn=a[1][1]
    
    fpr=fp/(fp+tn)
    fnr=fn/(fn+tn)
    ter=(fp+fn)/(tp+tn+fn+fp)
    fpr_NB.append(fpr)
    fnr_NB.append(fnr)
    ter_NB.append(ter)
    acc_NB.append(acc)

In [10]:
print("________________________________________NAIVE BAYES________________________________________________________________")
print("Folds\t\tFalse-Positive-Rate\tFalse Negative Rate\t \t Error Rate\t \tAccuracy")
print("________________________________________________________________________________________________________")
for i in range(0,10):
    print(i+1,"\t\t\t%.4f"%fpr_NB[i],"\t\t\t%.4f"%fnr_NB[i],"\t\t\t%.4f"%ter_NB[i],"\t\t\t%.4f"%acc_NB[i])
print("________________________________________________________________________________________________________")
print("Average\t\t\t",round(statistics.mean(fpr_NB),4),"\t\t",round(statistics.mean(fnr_NB),4),"\t\t",round(statistics.mean(ter_NB),4),"\t\t",round(statistics.mean(acc_NB),4))


________________________________________NAIVE BAYES________________________________________________________________
Folds		False-Positive-Rate	False Negative Rate	 	 Error Rate	 	Accuracy
________________________________________________________________________________________________________
1 			0.2708 			0.0385 			0.1562 			0.8438
2 			0.2552 			0.0220 			0.1410 			0.8590
3 			0.2193 			0.0220 			0.1171 			0.8829
4 			0.2489 			0.0166 			0.1348 			0.8652
5 			0.2009 			0.0552 			0.1152 			0.8848
6 			0.3008 			0.0110 			0.1717 			0.8283
7 			0.2903 			0.0276 			0.1674 			0.8326
8 			0.2273 			0.0608 			0.1326 			0.8674
9 			0.4801 			0.0608 			0.3660 			0.6340
10 			0.4030 			0.1160 			0.2810 			0.7190
________________________________________________________________________________________________________
Average			 0.2897 		 0.043 		 0.1783 		 0.8217


In [13]:
fpr_DT=[]
fnr_DT=[]
ter_DT=[]
acc_DT=[]
for train_index, test_index in skf.split(X,y):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y[train_index], y[test_index]
    
    
    clf_DT.fit(X_train, y_train)
    pred = clf_DT.predict(X_test)
    
    acc = round(accuracy_score(y_test, pred),4)
    a=confusion_matrix(y_test, pred)
    tp=a[0][0]
    fp=a[0][1]
    fn=a[1][0]
    tn=a[1][1]
    
    fpr=fp/(fp+tn)
    fnr=fn/(fn+tn)
    ter=(fp+fn)/(tp+tn+fn+fp)
    fpr_DT.append(fpr)
    fnr_DT.append(fnr)
    ter_DT.append(ter)
    acc_DT.append(acc)

In [14]:
print("________________________________________Decision Trees________________________________________________________________")
print("Folds\t\tFalse-Positive-Rate\tFalse Negative Rate\t \t Error Rate\t \tAccuracy")
print("________________________________________________________________________________________________________")
for i in range(0,10):
    print(i+1,"\t\t\t%.4f"%fpr_DT[i],"\t\t\t%.4f"%fnr_DT[i],"\t\t\t%.4f"%ter_DT[i],"\t\t\t%.4f"%acc_DT[i])
print("________________________________________________________________________________________________________")
print("Average\t\t\t",round(statistics.mean(fpr_DT),4),"\t\t",round(statistics.mean(fnr_DT),4),"\t\t",round(statistics.mean(ter_DT),4),"\t\t",round(statistics.mean(acc_DT),4))


________________________________________Decision Trees________________________________________________________________
Folds		False-Positive-Rate	False Negative Rate	 	 Error Rate	 	Accuracy
________________________________________________________________________________________________________
1 			0.0872 			0.1374 			0.0868 			0.9132
2 			0.0698 			0.1209 			0.0738 			0.9262
3 			0.1130 			0.1374 			0.0976 			0.9024
4 			0.0920 			0.1271 			0.0848 			0.9152
5 			0.0939 			0.0939 			0.0739 			0.9261
6 			0.0934 			0.0884 			0.0717 			0.9283
7 			0.0370 			0.1381 			0.0674 			0.9326
8 			0.1443 			0.0829 			0.0935 			0.9065
9 			0.2356 			0.1215 			0.1547 			0.8453
10 			0.2487 			0.2155 			0.1874 			0.8126
________________________________________________________________________________________________________
Average			 0.1215 		 0.1263 		 0.0991 		 0.9008


In [16]:
fpr_LR=[]
fnr_LR=[]
ter_LR=[]
acc_LR=[]
for train_index, test_index in skf.split(X,y):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y[train_index], y[test_index]
    
    clf_LR.fit(X_train, y_train)
    
    pred = clf_LR.predict(X_test)
    
    acc = round(accuracy_score(y_test, pred),4)
    a=confusion_matrix(y_test, pred)
    tp=a[0][0]
    fp=a[0][1]
    fn=a[1][0]
    tn=a[1][1]
    
    fpr=fp/(fp+tn)
    fnr=fn/(fn+tn)
    ter=(fp+fn)/(tp+tn+fn+fp)
    fpr_LR.append(fpr)
    fnr_LR.append(fnr)
    ter_LR.append(ter)
    acc_LR.append(acc)

In [17]:
print("________________________________________Logistic Regression________________________________________________________________")
print("Folds\t\tFalse-Positive-Rate\tFalse Negative Rate\t \t Error Rate\t \tAccuracy")
print("________________________________________________________________________________________________________")
for i in range(0,10):
    print(i+1,"\t\t\t%.4f"%fpr_LR[i],"\t\t\t%.4f"%fnr_LR[i],"\t\t\t%.4f"%ter_LR[i],"\t\t\t%.4f"%acc_LR[i])
print("________________________________________________________________________________________________________")
print("Average\t\t\t",round(statistics.mean(fpr_LR),4),"\t\t",round(statistics.mean(fnr_LR),4),"\t\t",round(statistics.mean(ter_LR),4),"\t\t",round(statistics.mean(acc_LR),4))


________________________________________Logistic Regression________________________________________________________________
Folds		False-Positive-Rate	False Negative Rate	 	 Error Rate	 	Accuracy
________________________________________________________________________________________________________
1 			0.0482 			0.1319 			0.0694 			0.9306
2 			0.0651 			0.1319 			0.0759 			0.9241
3 			0.0909 			0.1209 			0.0824 			0.9176
4 			0.0667 			0.0718 			0.0543 			0.9457
5 			0.0955 			0.1105 			0.0804 			0.9196
6 			0.1094 			0.0552 			0.0674 			0.9326
7 			0.0237 			0.0884 			0.0435 			0.9565
8 			0.0678 			0.0884 			0.0609 			0.9391
9 			0.2172 			0.1436 			0.1503 			0.8497
10 			0.1676 			0.2044 			0.1438 			0.8562
________________________________________________________________________________________________________
Average			 0.0952 		 0.1147 		 0.0828 		 0.9172


In [19]:
fpr_RF=[]
fnr_RF=[]
ter_RF=[]
acc_RF=[]
from sklearn.ensemble import RandomForestClassifier
clf_RF = RandomForestClassifier(n_estimators=100)
for train_index, test_index in skf.split(X,y):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y[train_index], y[test_index]
    
    
    clf_RF.fit(X_train, y_train)
    pred = clf_RF.predict(X_test)
    
    acc = round(accuracy_score(y_test, pred),4)
    a=confusion_matrix(y_test, pred)
    tp=a[0][0]
    fp=a[0][1]
    fn=a[1][0]
    tn=a[1][1]
    
    fpr=fp/(fp+tn)
    fnr=fn/(fn+tn)
    ter=(fp+fn)/(tp+tn+fn+fp)
    fpr_RF.append(fpr)
    fnr_RF.append(fnr)
    ter_RF.append(ter)
    acc_RF.append(acc)

In [20]:
print("________________________________________Random Forest________________________________________________________________")
print("Folds\t\tFalse-Positive-Rate\tFalse Negative Rate\t \t Error Rate\t \tAccuracy")
print("________________________________________________________________________________________________________")
for i in range(0,10):
    print(i+1,"\t\t\t%.4f"%fpr_RF[i],"\t\t\t%.4f"%fnr_RF[i],"\t\t\t%.4f"%ter_RF[i],"\t\t\t%.4f"%acc_RF[i])
print("________________________________________________________________________________________________________")
print("Average\t\t\t",round(statistics.mean(fpr_RF),4),"\t\t",round(statistics.mean(fnr_RF),4),"\t\t",round(statistics.mean(ter_RF),4),"\t\t",round(statistics.mean(acc_RF),4))


________________________________________Random Forest________________________________________________________________
Folds		False-Positive-Rate	False Negative Rate	 	 Error Rate	 	Accuracy
________________________________________________________________________________________________________
1 			0.0400 			0.0769 			0.0456 			0.9544
2 			0.0511 			0.0824 			0.0521 			0.9479
3 			0.0366 			0.1319 			0.0651 			0.9349
4 			0.0355 			0.0994 			0.0522 			0.9478
5 			0.0452 			0.0663 			0.0435 			0.9565
6 			0.0492 			0.0387 			0.0348 			0.9652
7 			0.0229 			0.0552 			0.0304 			0.9696
8 			0.0227 			0.0497 			0.0283 			0.9717
9 			0.1854 			0.0773 			0.1133 			0.8867
10 			0.2011 			0.1878 			0.1547 			0.8453
________________________________________________________________________________________________________
Average			 0.069 		 0.0866 		 0.062 		 0.938


SUMMARY
------------------------------------------------------------------------------


Tried Spam/Non-Spam Email classfication with Below Classifiers 
1. Multinomial Naive Bayes,
2. Decision Trees, 
3. Logistic Regression 
4. Random Forest 
---------------------------------------------------------------------------------------------------------------------------------------

Best ERROR Rate, False-Positive Rate and False Negative Rate is given by RANDOM FOREST
--------------------------------------------------------------------------------------------
Average False Positve Rate   0.069
--------------------------------------------------------------------------------------------
Average False Negative Rate  0.0866
--------------------------------------------------------------------------------------------
Average Error Rate           0.062
--------------------------------------------------------------------------------------------
Average Accuracy             0.938
--------------------------------------------------------------------------------------------





